### PR TITLE
[AST] Add Builtin.packLength

### PR DESCRIPTION
**Pull request by Azoy**
Add a new builtin which takes a pack expansion and counts the number of items within it. This is built on top of https://github.com/apple/swift/pull/66638.

### CODE REVIEW

File: lib/AST/Builtins.cpp [hunk 1]
```diff
@@ -1932,6 +1945,19 @@ static ValueDecl *getHopToActor(ASTContext &ctx, Identifier id) {
   return builder.build(id);
 }
 
+static ValueDecl *getPackCount(ASTContext &ctx, Identifier id) {
+  BuiltinFunctionBuilder builder(ctx, /* genericParamCount */ 1,
+                                 /* anyObject */ false,
+                                 /* areParametersPack */ true);
+
+  auto packEx = makeGenericParam(/* index */ 0, /* isPackExpansion */ true);
```
Comment (slavapestov) [inline lib/AST/Builtins.cpp:1953]: It seems a bit odd to have an isPackExpansion flag here. Why not handle this like metatypes and tuples, and add a `makePackExpansion()` utility?

File: lib/AST/Builtins.cpp [hunk 2]
```diff
@@ -1932,6 +1945,19 @@ static ValueDecl *getHopToActor(ASTContext &ctx, Identifier id) {
   return builder.build(id);
 }
 
+static ValueDecl *getPackCount(ASTContext &ctx, Identifier id) {
+  BuiltinFunctionBuilder builder(ctx, /* genericParamCount */ 1,
+                                 /* anyObject */ false,
+                                 /* areParametersPack */ true);
```
Comment (slavapestov) [inline lib/AST/Builtins.cpp:1951]: areParameterPacks maybe not areParametersPack?

File: lib/IRGen/GenBuiltin.cpp [hunk 3]
```diff
@@ -1345,5 +1345,30 @@ void irgen::emitBuiltinCall(IRGenFunction &IGF, const BuiltinInfo &Builtin,
...
 
+  if (Builtin.ID == BuiltinValueKind::PackCount) {
+    (void) args.claimAll();
+
+    auto metaTy = argTypes[0].getASTType()->castTo<MetatypeType>();
+    auto tupleTy = metaTy->getInstanceType()->castTo<TupleType>();
+
+    llvm::Value *count = llvm::ConstantInt::get(IGF.IGM.SizeTy, 0);
+
+    // Accumulate the count of the pack.
```
Comment (slavapestov) [inline lib/IRGen/GenBuiltin.cpp:1356]: It seems odd how you have to iterate over the tuple elements here. Can't you get the PackType from the substitution map, in the same way that the SILCombinerBuiltinVisitors does?

Maybe the builtin should not take a tuple at all? A pack of metatypes should be better. You mentioned that this generated some dead instructions, but we could find a way around that.

**Review by slavapestov — COMMENTED** [review]
Cool! Just a few minor nitpicks.

**Comment by Azoy** [comment]
@swift-ci please test

File: lib/AST/Builtins.cpp [hunk 4]
```diff
@@ -767,6 +773,15 @@ namespace {
         return MetatypeType::get(Object.build(builder), Repr);
       }
     };
+    struct ParameterPackGenerator {
```
Comment (slavapestov) [inline lib/AST/Builtins.cpp:776]: This should be named PackExpansionGenerator

File: lib/AST/Builtins.cpp [hunk 5]
```diff
@@ -813,6 +828,11 @@ makeMetatype(const T &object, Optional<MetatypeRepresentation> repr = None) {
   return { object, repr };
 }
 
+static BuiltinFunctionBuilder::ParameterPackGenerator
+makeParameterPack(unsigned index = 0) {
```
Comment (slavapestov) [inline lib/AST/Builtins.cpp:832]: makePackExpansion(). The pattern can be any type, not just a generic parameter type, so it should work exactly like makeMetatype(), etc

File: lib/IRGen/GenBuiltin.cpp [hunk 6]
```diff
@@ -1345,5 +1345,15 @@ void irgen::emitBuiltinCall(IRGenFunction &IGF, const BuiltinInfo &Builtin,
     return;
   }
 
+  if (Builtin.ID == BuiltinValueKind::PackCount) {
```
Comment (slavapestov) [inline lib/IRGen/GenBuiltin.cpp:1348]: Can we lower away the builtin at the SIL level and turn it into a SIL `pack_count` instruction?

**Comment by Azoy** [comment]
@swift-ci please test

**Comment by Azoy** [comment]
@swift-ci please test

**Pull request merged**